lib/core/CStaticThreadPool.cc

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/CStaticThreadPool.h>
+
+namespace ml {
+namespace core {
+namespace {
+std::size_t computeSize(std::size_t hint) {
+    std::size_t bound{std::thread::hardware_concurrency()};
+    std::size_t size{bound > 0 ? std::min(hint, bound) : hint};
+    return std::max(size, std::size_t{1});
+}
+}
+
+CStaticThreadPool::CStaticThreadPool(std::size_t size)
+    : m_Cursor{0}, m_TaskQueues{computeSize(size)} {
+    m_Pool.reserve(m_TaskQueues.size());
+    for (std::size_t id = 0; id < m_TaskQueues.size(); ++id) {
+        try {
+            m_Pool.emplace_back([&, id] { worker(id); });
+        } catch (...) {
+            this->shutdown();
+            throw;
+        }
+    }
+}
+
+CStaticThreadPool::~CStaticThreadPool() {
+    this->shutdown();
+}
+
+void CStaticThreadPool::schedule(TTask&& task) {
+    m_TaskQueues[m_Cursor.fetch_add(1) % m_TaskQueues.size()].push(std::forward<TTask>(task));
+}
+
+void CStaticThreadPool::schedule(std::function<void()>&& f) {
+    this->schedule(TTask{[g = std::move(f)] {
+        g();
+        return boost::any{};
+}
+});
+}
+
+void CStaticThreadPool::shutdown() {
+    // Signal to each thread that it is finished.
+    for (auto& queue : m_TaskQueues) {
+        queue.push(TTask{[&] {
+            m_Done = true;
+            return boost::any{};
+        }});
+    }
+    for (auto& thread : m_Pool) {
+        thread.join();
+    }
+}
+
+void CStaticThreadPool::worker(std::size_t id) {
+
+    auto noThrowExecute = [](TOptionalTask& task) {
+        try {
+            (*task)();
+        } catch (const std::future_error& e) {
+            LOG_ERROR(<< "Failed executing packaged task: " << e.code() << " "
+                      << e.what());
```

Comment:
if you get a task failure like this: what does it mean for the overall process? Except from the logline, I see no error being populated back. Maybe `schedule(...)` could take a callback parameter which will be executed in an error situation?

Reply:
Basically, I wanted to avoid this causing the worker thread to exit. I mean this really shouldn't happen, I was being paranoid because the tasks are passed in. I think in both cases it throws: task already executed (the thread pool really should ensure this doesn't happen if being used in a sane way) and not initialised the right behaviour is to just ignore, but felt uncomfortable silently swallowing an exception.

This goes to our discussions earlier about overall monitoring strategy, because I think it is unlikely the calling code can act on this failure, it really means that there was probably a logic failure somewhere in the program. Having a variable which has to be dealt with in every calling context creates a lot of debt.